rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+
+In synchronous training, all replicas are in sync with each other and gradients
+are aggregated across all the replicas. In fully asynchronous training, all
+replicas operate asynchronously and update the variables independently. We can
+also support a hybrid approach, where the replicas are divided into "sync
+groups" - within each sync group, each replica is in sync, but the sync groups
+are asynchronous among themselves. To support this, we later define
+"num_replicas" as the total number of replicas (sync or async) and
+"num_replicas_in_sync" as the number of replicas in one sync group.
+
+#### Input Pipeline
+
+An instance of the input pipeline (e.g. `Dataset`, `FIFOQueue`) obtained by
+calling the user's `input_fn` once. One input pipeline may be used to feed one
+or more model replicas (on one or more workers). See
+[InputReplicationMode](#inputreplicationmode).
+
+### Replicator API
+
+#### Detailed Design
+
+Here are the main ingredients of the API:
+
+*   `Replicator` class: This is the main class of the API and provides the main
+    methods to replicate the computation.
+    *   `__init__:`Creates a Replicator instance. It takes a
+        `DistributionStrategy` object which is used for (almost) all the
+        underlying operations such as replicating the computation, all-reduce
+        etc.
+    *   `scope`: A context manager that manages variable creation correctly
+        within its scope, and also sets the distribution strategy as current for
+        the current graph & thread. Variables that need to be replicated must be
+        created within `Replicator.scope`, unless they are created inside the
+        `fn` passed to `run` which automatically wraps the entire `fn` inside
+        `Replicator.scope`. Entering a `scope` twice is safe, so in general,
+        enclosing everything that is potentially distributed inside a `scope`
+        should be safe.
+    *   `run`: This method is used to replicate the computation. It takes a
+        function `fn` which specifies the computation, and an (optional)
+        `InputIterator` `inputs` to be fed into`fn`. It calls `fn` with one
+        element from the input iterator within distribution strategy `scope`.
+        Variables and optimizers for the model can be created inside or outside
+        `fn`, as long as they're created within `scope`. `run` can be called
+        multiple times with different `fn`. In graph mode, run simply creates
+        the replicated graph whereas in eager mode, it will run the actual
+        computation. It then returns the list of outputs of `fn` from each
+        replica.
+    *   `prepare_input`: This method is used to prepare the input for being
+        passed to run. It takes an `InputFn` `fn` which either returns a
+        `Dataset` or a function which returns
+        ([nests](https://www.tensorflow.org/api_docs/python/tf/contrib/framework/nest)
+        of) input tensors. It also takes an optional `InputReplicationMode`
```

Comment:
Do we need to support both Dataset and nests of tensors? Can't a dataset of nested tensors suffice for the latter use case, such that this always returns a dataset?

Reply:
I recall, perhaps once upon a time, that there was a restriction that the Estimator input_fn had to return datasets to work with DistStrat, and that returning a tuple of (feature, label) would not be supported. Or maybe that was with replicate_model_fn? Does that restriction still hold, or does the acceptance of nested tensors imply that other return types are also acceptable?

Reply:
Great point. I think once we add support for fn returning tensors, we should be able to support `input_fn` returning (feature,label) tuples in Estimator as well.